Lib/zipimport.py

```diff
@@ -321,6 +321,16 @@ def get_resource_reader(self, fullname):
         return ZipReader(self, fullname)
 
 
+    def invalidate_caches(self):
+        """Reload the file data of the archive path."""
+        try:
+            self._files = _read_directory(self.archive)
```

Comment:
Thanks, @desmondcheongzx. I do appreciate this fix.

Just dropping a comment in case the impact of this was missed during the review.
This might impact the performance notably. For example, I only import two zipped packages `py4j` and `pyspark` but it slows down the speed of `importlib.invalidate_caches` up to 3500% because now it has to go through the zipped package.

Reply:
This was missed unfortunately.

To get everyone back up to speed, in the [original discussion](https://github.com/python/cpython/pull/24159#discussion_r571303809) there were two decisions made:
1.  The `_get_files()` method should be as fast as possible in the common case where zip files are not changed. This is why we avoided a stat call on every access of the importer.
2. Rebuild the cache when `invalidate_caches()` is called. This second decision might not be the right one, because clearing a cache does not mean that we want to immediately repopulate it.

There are three relevant events:
1. `_get_files()` is called with a valid cache
2. `invalidate_caches` is called
3. `_get_files()` is called with an __invalid__ cache

IMO, what we really want is for events (1) and (2) to be fast, while event (3) can be slow since we're repopulating a cache. (Note that in the original PR we made (1) and (3) fast, but (2) slow). So I propose the following:

- Add a boolean flag `cache_is_valid` that is set to false when `invalidate_caches()` is called.
- In `_get_files()`, if `cache_is_valid` is true, use the cache. If `valid_cache` is false, call `_read_directory()`.

This approach avoids any behaviour change and keeps the common path performant, while also shifting the cost of reading the directory out of cache invalidation.

If we want to go further and consider the fact that we rarely expect zip archives to change, then we can also consider adding a flag to `importlib.invalidate_caches` so that users can choose whether zip caches are invalidated.

Any thoughts, @brettcannon?